support-models/src/main/scala/com/gu/support/workers/states/CreateZuoraSubscriptionState.scala

```diff
@@ -1,31 +1,100 @@
   acquisitionData: Option[AcquisitionData]
 ) extends FailureHandlerState
 
-import com.gu.support.encoding.Codec
-import com.gu.support.encoding.Codec._
+object CreateZuoraSubscriptionWrapperState {
+  implicit val codec = deriveCodec[CreateZuoraSubscriptionWrapperState]
+}
+
+sealed trait CreateZuoraSubscriptionState
 
 object CreateZuoraSubscriptionState {
-  import com.gu.support.encoding.CustomCodecs._
-  implicit val codec: Codec[CreateZuoraSubscriptionState] = deriveCodec
+
+  case class CreateZuoraSubscriptionContributionState(
+    product: Contribution,
+    paymentMethod: PaymentMethod,
+    salesForceContact: SalesforceContactRecord,
+  ) extends CreateZuoraSubscriptionState
+
+  case class CreateZuoraSubscriptionDigitalSubscriptionDirectPurchaseState(
+    billingCountry: Country,
+    product: DigitalPack,
+    paymentMethod: PaymentMethod,
+    promoCode: Option[PromoCode],
+    salesForceContact: SalesforceContactRecord,
+  ) extends CreateZuoraSubscriptionState
+
+  case class CreateZuoraSubscriptionDigitalSubscriptionGiftPurchaseState(
+    billingCountry: Country,
+    giftRecipient: DigitalSubscriptionGiftRecipient,
+    product: DigitalPack,
+    paymentMethod: PaymentMethod,
+    promoCode: Option[PromoCode],
+    salesforceContacts: SalesforceContactRecords,
+  ) extends CreateZuoraSubscriptionState
+
+  case class CreateZuoraSubscriptionDigitalSubscriptionCorporateRedemptionState(
+    product: DigitalPack,
+    redemptionData: RedemptionData,
+    salesForceContact: SalesforceContactRecord,
+  ) extends CreateZuoraSubscriptionState
+
+  case class CreateZuoraSubscriptionDigitalSubscriptionGiftRedemptionState(
+    userId: String,
+    product: DigitalPack,
+    redemptionData: RedemptionData,
+  ) extends CreateZuoraSubscriptionState
+
+  case class CreateZuoraSubscriptionPaperState(
+    user: User,
+    product: Paper,
+    paymentMethod: PaymentMethod,
+    firstDeliveryDate: LocalDate,
+    promoCode: Option[PromoCode],
+    salesForceContact: SalesforceContactRecord,
+  ) extends CreateZuoraSubscriptionState
+
+  case class CreateZuoraSubscriptionGuardianWeeklyState(
+    user: User,
+    giftRecipient: Option[WeeklyGiftRecipient],
+    product: GuardianWeekly,
+    paymentMethod: PaymentMethod,
+    firstDeliveryDate: LocalDate,
+    promoCode: Option[PromoCode],
+    salesforceContacts: SalesforceContactRecords,
+  ) extends CreateZuoraSubscriptionState
+
+  private val discriminatedType = new DiscriminatedType[CreateZuoraSubscriptionState]("productType")
+  implicit val codec: Codec[CreateZuoraSubscriptionState] = discriminatedType.codec(List(
+    discriminatedType.variant[CreateZuoraSubscriptionContributionState]("Contribution"),
+    discriminatedType.variant[CreateZuoraSubscriptionDigitalSubscriptionDirectPurchaseState]("DigitalSubscriptionDirectPurchase"),
+    discriminatedType.variant[CreateZuoraSubscriptionDigitalSubscriptionGiftPurchaseState]("DigitalSubscriptionGiftPurchase"),
```

Comment:
These discriminator strings could be shared as vals between this class and `SendThankYouEmailState`

Reply:
there's no technical reason why they need to be the same, but I can see we want consistency so I will do that.